### PR TITLE
Error returning

### PR DESCRIPTION
**Pull request by erikpartila**
Hey,
I started using your npm package today for my custom built API. I want to be able to return a somewhat meaningful error message to the user of my API whenever there is an error, but with your implementation, the error only gets logged into the console and there is no way of retrieving it. I did some changes to implement this. All you need is pass a option paramater when creating new ZapClient. I also added an option to enable/disable error logging in console.

Would it be pease possible to merge these changes so I can use these new features ? Thanks in advance !

Erik

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,6 +16,7 @@
 </a>
 
 <br/><br/>
+
```
Comment (thc202) [inline README.md:19]: Revert unnecessary changes.

**Comment by thc202** [comment]
That we didn't return the error was/is an oversight, I don't think we should be adding options to control it though.

File: README.md (hunk 1 above)
Comment (erikpartila) [reply README.md]: done

Comment (thc202) [reply README.md]: There's still a change that should be reverted.

**Comment by thc202** [comment]
There shouldn't be any sensitive data, anything returned by the API is already known to the user (assuming the API is not unsecured).

Better to return a failed promise either way (I don't mind which object is returned though, as long as it includes the original response body).

**Comment by thc202** [comment]
The changelog should be updated to mention the fix.

**Comment by erikpartila** [comment]
@thc202 I updated the PR, hope everything checks out !

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ## [Unreleased]
 ### Changed
 * Update the link to API docs in README.md
+### Added
+* Possibilty to turn on/off error logging and return of error whenever request to OWASP ZAP fails.
```
Comment (thc202) [inline CHANGELOG.md:10]:
```suggestion
### Fixed
* Return errors (e.g. connection, ZAP API) with a rejected promise.
```

File: README.md [hunk 3]
```diff
@@ -43,6 +43,8 @@ const zapOptions = {
     host: '127.0.0.1',
     port: 8080,
   },
+  logError: true,
+  returnError: true,
```
Comment (thc202) [inline README.md:47]:
```suggestion
```

File: src/index.js [hunk 4]
```diff
@@ -66,7 +66,9 @@ function ClientApi (options) {
     params: {},
     baseURL: BASE_URL_JSON,
     headers: options.apiKey ? { 'X-ZAP-API-Key': options.apiKey } : {},
-    proxy: options.proxy
+    proxy: options.proxy,
+    logError: options.logError,
+    returnError: options.returnError
```
Comment (thc202) [inline src/index.js:71]:
```suggestion
    proxy: options.proxy
```

File: src/index.js [hunk 5]
```diff
@@ -122,7 +124,16 @@ ClientApi.prototype.request = async (url, data, format) => {
...
+    const requestConfig = structuredClone(defaultAxiosConfig)
+    if (requestConfig.logError) {
+      console.log(error.message, ', Data:', error.response?.data)
+    }
+    if (requestConfig.returnError) {
+      return {
+        status: error.response.status,
+        data: error.response.data
+      }
+    }
```
Comment (thc202) [inline src/index.js:136]:
```suggestion
    return Promise.reject(new ApiClientError("Error during request.", error))
```

**Comment by thc202** [comment]
Note that I did not add the `ApiClientError` but it's straightforward error, which should have the object with the error details. The original error might not have a `status` so that should be handled properly (e.g. when failed to connect to ZAP).

The readme should mention what the error has.

File: src/index.js [hunk 6]
```diff
@@ -109,6 +109,14 @@ function ClientApi (options) {
   this.websocket = new Websocket(this)
 }
 
+class ApiClientError extends Error {
+  constructor(message, originalError) {
```
Comment (thc202) [inline src/index.js:113]:
```suggestion
  constructor (message, originalError) {
```

File: src/index.js [hunk 7]
```diff
@@ -122,7 +130,7 @@ ClientApi.prototype.request = async (url, data, format) => {
 
     return response.data
   } catch (error) {
-    console.log(error.message, ', Data:', error.response?.data)
+    return Promise.reject(new ApiClientError("Error during request.", error))
```
Comment (thc202) [inline src/index.js:133]:
```suggestion
    return Promise.reject(new ApiClientError('Error during request.', error))
```

**Comment by thc202** [comment]
@erikpartila please check the latest changes, the original error is now being exposed through  the `cause` (more idiomatic) and the response details are exposed as the initial changeset (for the user to not rely on internal details).

**Review by psiinon — APPROVED** [review]

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!
